### ISSUE TITLE
is_action

### ISSUE THREAD

**Issue by fenhl** [museun/twitchchat#120]
I'm upgrading from 0.7 to 0.10 and it looks like [`twitchchat::messages::Privmsg`](https://docs.rs/twitchchat/0.10.3/twitchchat/messages/struct.Privmsg.html) no longer has an [`is_action`](https://docs.rs/twitchchat/0.7.2/twitchchat/commands/struct.PrivMsg.html#method.is_action) method. Is there another way to check whether a message is a `/me` action?

**Comment by museun** [comment]
Ah. an oversight. Actions are actually 'CTCP' extensions of the IRC protocol. The way its done is by sending a non-printable byte (`0x01`) before and after the data segment of the message.

You could check the data field to see if it started/ended with this byte. I'll add the method in the next release (tomorrow probably).

Ideally I should strip those two bytes out and expose an actions fields/method like the old version did.

I completely forgot that Twitch allowed a limited form of CTCP.

**Comment by museun** [comment]
Err. a correction. Its in the form of `\x01COMMAND data\x01\r\n`. The CTCP 'command' follows the first `\x01` byte followed by a space (`\x20`), then the rest of the data payload, then the final `\x01` byte.  So, `COMMAND` would be `ACTION` here. 
My crate strips the `\r\n` off when doing parsing. 

So, something like 
```rust
// assuming you've determined it has the marker bytes
let mut iter = splitn(2, data[1,..data.len()-1]);
match iter.next() {
    // Get the reamining data
    Some("ACTION") => data.next().ok_or_else(|| EmptyData),
    // maybe handle this differently, but this wasn't an ACTION
    Some(unknown) => { Err(UnknownCommand(unknown.to_string())) }
    // Invalid data was passed (This should never happen. but if you're constructing your own data to simulate a connection who knows..)
    None => { Err(InvalidMessage) }
}
```

**Comment by museun** [comment]
And a final update, I think I'll have an `is_action() -> bool` and `is_ctcp() -> Option<PrivmsgCtcp>` where ` #[non_exhaustive] enum PrivmsgCtcp { Action, Unknown<Cow<'t, str> }`. 
Just incase there are other ones. I can atleast expose them and when I'm notified that others do exist, I can add them to the enum without breaking things.

**Comment by fenhl** [comment]
Perhaps it would make sense to have a convenience method for extracting the content of the action, i.e. without the surrounding `\x01ACTION ` and `\x01`. Then `msg.is_action()` could also just be `msg.action().is_some()`.

**Comment by Emilgardis** [comment]
I like that design choice of returning an instance of `Option<Ctcp>`

**Comment by museun** [comment]
> 
> 
> Perhaps it would make sense to have a convenience method for extracting the content of the action, i.e. without the surrounding `\x01ACTION ` and `\x01`. Then `msg.is_action()` could also just be `msg.action().is_some()`.

It would strip out the CTCP (client-to-client protocol) and a flag (`is_action` / `is_ctcp`) would be provided.
These kinds of 'CTCP' are essentially just `PRIVMSG`s.

So, the parser would handle that and set the flag accordingly. the `data` field would be unchanged (sans the marker bytes and the 'command' stripped).



> 
> 
> I like that design choice of returning an instance of `Option<Ctcp>`

Should 'Ctcp' just be a simple enum like I described above? Or should it also duplicate the 'data' section? i.e.:

```rust
#[non_exhaustive]
enum Ctcp<'t> {
    Action { data: Cow<'t, str> },
    Unknown { command: Cow<'t, str>, data: Cow<'t, str> }
}
```

**Comment by Emilgardis** [comment]
I have no strong preference, duplicating seems fair but does make it larger for the Owned case, so I'd say keep data in Privmsg

**Comment by museun** [comment]
I implemented it in 758933cb79a0981b513012e61149a06506f4c347

**Comment by museun** [comment]
These changes will be in `v0.11.0`. I've changed the API enough that it breaks semver. Notably replacing some `&Cow` with `Cow` and moving the `connect` module around for the future update to de-tokioify the library.

It should be out later today, I need to redo the run-loop for #119
